design-proposals/storage-class-migration/storage-class-migration.md

```diff
@@ -0,0 +1,276 @@
+# Overview
+This design proposal extends KubeVirt APIs to enable the migration between PVCs or storage classes.
+
+## Motivation
+In some circumstances, the guest owner may want to have the option of switching the storage provider or class for a group of volumes used by an active VM. The VM must currently be turned off and recreated for this operation causing service disruptions and downtime.
+
+
+## Goals
+Design and develop an API to handle storage migrations.
+
+ ## Non  Goals
+*  This features is not directly aimed for live migrating the VM using RWO
+   volumes. Hence, the API covers and refers only to the storage migration.
+
+ ## Definition of Users
+ This feature addresses for all KubeVirt users who want to move to another storage classes or PVCs without workload interuptions.
+
+## User Stories
+  1. A storage class of volumes used by a running VM is deprecated
+  2. A more performant storage class is now available in the cluster and the user wants start using it.
+  3. The storage device is filling up
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD `StorageMigration`
+
+```golang
+type StorageMigration struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec StorageMigrationSpec `json:"spec"`
+        Status *StorageMigrationStatus `json:"status,omitempty"`
+}
+
+// StorageMigrationList is a list of StorageMigration resources
+type StorageMigrationList struct {
+        metav1.TypeMeta `json:",inline"`
+        metav1.ListMeta `json:"metadata"`
+        Items []StorageMigration `json:"items"`
+}
+
+type MigratedVolume struct {
+        SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+        DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+}
+
+type MigrationStorageClass struct {
+        SourceStorageClass      string `json:"sourceStorageClass,omitempty" valid:"required"`
+        DestinationStorageClass string `json:"destinationStorageClass,omitempty" valid:"required"`
+        // PrefixDestinationVolume is the prefix used by KubeVirt to generate
+        // the destination PVC. The claim name will contain the prefix plus a
+        // random string.
+        PrefixDestinationVolume string `json:"PrefixDestinationVolume,omitempty" valid:"required"`
+}
+
+type ReclaimPolicySourcePvc string
+
+const (
+        DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+)
+
+// StorageMigrationSpec is the spec for a StorageMigration resource
+type StorageMigrationSpec struct {
+        // The name of the VMI to perform the migration on. VMI must exist in the storage migration objects
+        // namespace
+        VMIName string `json:"vmiName,omitempty" valid:"required"`
+        // MigrationVolumes is a list of volumes to migrate
+        MigrationVolumes []MigrationVolume `json:"migrationVolumes,omitempty"`
```

Comment:
What is the reason we are separating the MigrationVolume from the MigrationStorageClass? To me the two are tightly linked. In theory each volume could be migrated to a different storage class. So you would have a triple of `<source volume, target volume (generated if not specified), target storageclass>`. Maybe even the retain policy as well.

How would you map the migration target storage classes properly otherwise?

Reply:
@awels are a different specification and usage IMO. 

In MigrationVolumes, KubeVirt doesn't need to create anything, it expects to use the `DestinationPvc`. For the `MigrationStorageClass ` KubeVirt takes care of creating the PVC object. 
I think this distinction can be useful when for example the user wants to use a different size or access mode. For this case, I would expect that they will create the PVC manually and reference it in the `DestinationPVC`.
If they specify only the destination storage class, KubeVirt would need to infer the PVC specification for the original one (e.g size and access mode).

Does this make sense to you?

Reply:
Additionally, the idea of keeping the 2 separates is for the cases when the users want to only specify the storage class. Like I want to migrate any volumes of `storage class A` into `storage class B`.

We might want to improve the PVC renaming right now wit the prefix is a bit awkward, but for example, we could also define policy renaming like preserving the original name of the PVC and adding `-copy` in the end, or a random string.